src/batcher/solve_pair.rs

```diff
@@ -1,16 +1,435 @@
 use crate::models::order::Order;
 use crate::models::solution::Solution;
 use anyhow::Result;
-use web3::types::U256;
+use web3::types::{U256, H256, Address};
+
+
+pub enum OrderPairType {
+    LhsFullyFilled,
+    RhsFullyFilled,
+    BothFullyFilled,
+}
+
+
+trait Matchable {
+    /// Returns which of the two orders can be fully matched, or both, or none.
+    fn match_compare(&self, other: &Order) -> Option<OrderPairType>;
+    /// Returns whether the orders can be matched. For this the tokens need
+    /// to match, there must be a price that satisfies both orders.
+    fn attracts(&self, other: &Order) -> bool;
+    /// Returns whether this order's sell token is the other order's buy token
+    /// and vice versa.
+    fn opposite_tokens(&self, other: &Order) -> bool;
+    /// Returns whether there is a price that satisfies both orders.
+    fn have_price_overlap(&self, other: &Order) -> bool;
+}
+
+
+impl Matchable for Order {
+    fn match_compare(&self, other: &Order) -> Option<OrderPairType> {
+        if !self.attracts(other) {
+            return None;
+        }
+
+        if self.buy_amount <= other.sell_amount && self.sell_amount <= other.buy_amount {
+            Some(OrderPairType::LhsFullyFilled)
+        } else if self.buy_amount >= other.sell_amount && self.sell_amount >= other.buy_amount {
+            Some(OrderPairType::RhsFullyFilled)
+        } else {
+            Some(OrderPairType::BothFullyFilled)
+        }
+    }
+
+    fn attracts(&self, other: &Order) -> bool {
+        self.opposite_tokens(other) && self.have_price_overlap(other)
+    }
+
+    fn opposite_tokens(&self, other: &Order) -> bool {
```

Comment:
I think we can also assume that all orders are opposite_tokens if they are passed to the solver

Reply:
Removed quite some code that doesn't appear to be required for this first version in 096afc3 - also related to your previous comment:
> I am wondering whether we need this particular enum:
Since we will fill both orders fully, it is probably not really needed